### ISSUE TITLE
MicroProfile Context Propagation introduction

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/docs#473]
We need to provide introductory conceptual information about context propagation (concurrency). The new MP Context Propagation topic proposed by @njr-11 in #471 seems to assume that the reader has a lot of knowledge already.

We have the basis of a concept topic that we should be able to use (subject to confirmation by @njr-11 ) from WASdev.net. I've moved that topic into .adoc format already here https://github.com/lauracowen/blogs/blob/concurrency/drafts/concurrency.adoc but it needs updating and refocusing a bit so that MP Context Propagation and developing microservices for the cloud is its primary focus (but referring to Java SE and EE concurrency for context).

So, in summary, write a concept topic based on the .adoc file mentioned above to introduce developers to the basic ideas of context propagation and concurrency. Then ensure that the new topic proposed links back to this topic.

**Comment by njr-11** [comment]
Proposed edits to concurrency.adoc, highlighted in **_bold italics_**, with any notes about removed parts in [brackets].

Concurrency is being able to run multiple tasks in parallel, which can increase the efficiency of an application.  _**Tasks can be submitted to run immediately, at a specified time, or in response to the completion of one or more actions upon which they depend. [replaces subsequent sentence which was Java EE-specific]**_

**_For example, application code that computes the total cost of an order might be able to benefit by looking up and computing the state and local sales tax rates in one task, the shipping costs in another, and finally combining the results._**

**_[no changes to Concurrency in Java SE and Java EE, but insert the following section after it]_**

**_Concurrency in MicroProfile_**

**_MicroProfile applications have many of the same needs, but with a greater focus on reactive patterns.  One form of reactive programming is the Java SE [CompletionStage](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletionStage.html) model, which emphasizes creating pipelines of dependent actions that run upon completion of the states they depend.  In Java SE, these pipelines of dependent stages run with inconsistent thread context.  The [MicroProfile Context Propagation specification](https://download.eclipse.org/microprofile/microprofile-context-propagation-1.0/microprofile-context-propagation.html) solves this problem by introducing the [ManagedExecutor](https://openliberty.io/docs/ref/javadocs/microprofile-3.0-javadoc/org/eclipse/microprofile/context/ManagedExecutor.html), which is nearly identical to and fully compatible with the [ManagedExecutorService](https://openliberty.io/docs/ref/javadocs/liberty-javaee8-javadoc/javax/enterprise/concurrent/ManagedExecutorService.html) from Java EE, but is enhanced with the ability to create context-aware implementations of [CompletionStage](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletionStage.html)._**

...

Open Liberty takes this a step further because the default managed executors, all configured managed executors, **_including those created via MicroProfile Context Propagation,_** and the core server implementation all share a single common thread pool. This thread pool is automatically tuned throughout the lifetime of the server by measuring the impact of incrementally increasing and decreasing the number of threads.

Commonalities between Java SE and Java EE concurrency **_and MicroProfile_**

The Concurrency Utilities for Java EE specification **_and MicroProfile Context Propagation specification_** closely match**_[es -- remove plural here]_** the Concurrency Utilities specification included in Java SE, so much so that it is often possible to write applications using only the Java SE APIs but still taking advantage of Concurrency Utilities for Java EE **_and MicroProfile Context Propagation_**. Liberty further enables this by allowing you to inject executor services and thread factories under the Java SE interface names (for example, java.util.concurrent.ExecutorService) as well as the Java EE interface names (for example, javax.enterprise.concurrent.ManagedExecutorService).  **_With the [mpContextPropagation-1.0](https://openliberty.io/docs/ref/feature/#mpContextPropagation-1.0.html) feature enabled, Liberty also lets you cast these executors to the `org.eclipse.microprofile.context.ManagedExecutor` interface, which is useful whenever you require context-aware completion stages.  Once you obtain the completion stage from the managed executor, you interact with it exactly as you would the Java SE API, but with the ability to rely on consistent thread context in your completion stage actions._**

Thread context awareness

Another enhancement in Concurrency Utilities for Java EE and **_MicroProfile Context Propagation_** beyond Java SE is that executors are aware of thread context and they propagate the context of the task submitter to the thread of execution upon which the task **_or action_** runs. This allows tasks to perform lookups in the java:comp, java:module, or java:app namespace of the submitting thread, as well as run with the security context of the submitting thread, load classes from its thread context classloader, and so on. Liberty allows you to configure which types of thread context are propagated.

Thread factories are also aware of thread context and propagate the context of the thread that looked up or injected the thread factory to threads created by the thread factory. This has the effect of allowing thread pools that are created with these thread factories to create a pool of threads that all run under the same thread context; this largely avoids the costs of thread context switching. Managed executors and thread factories rely on another service provided by Concurrency Utilities for Java EE called the Context Service (javax.enterprise.concurrent.ContextService) for thread context propagation. Context Services are also available for direct usage, allowing you to achieve more flexible and fine grained control of thread context propagation. Context Services allow you to construct proxies that save the thread context at creation time and apply it upon invocation of interface methods.  **_In MicroProfile Context Propagation, the [ThreadContext](https://openliberty.io/docs/ref/javadocs/microprofile-3.0-javadoc/org/eclipse/microprofile/context/ThreadContext.html) API provides similar function._**

...

Try out the Liberty Concurrency **_and MicroProfile Context Propagation features_**

With the concurrent-1.0 feature, Liberty provides a robust, fully spec-compliant implementation of Concurrency Utilities for Java EE.  **_MicroProfile Context Propagation is available with the mpContextPropagation-1.0 feature.  Both features can be enabled and used in tandem._**  Try it out today, and start enjoying the advantages of parallelism and task scheduling in a Java EE environment.

**Comment by dmuelle** [comment]
Hi @lauracowen & @njr-11 
I'm posting my draft of the above content here, to make discussion easier. I,ve asked a few questions below. I'll post the next revision directly to the draft site.

After talking with Nathan, who confirmed that concurrency, rather than context propagation, should be the real focus of this doc, I edited the concurrency doc posted above. My main question is- should context propagation be discussed here? How to managed a focus on concurrency in the context of an API named Context Propagation?

= Managing concurrency in microservices

Concurrency is the ability to run multiple tasks at the same time, which can increase the efficiency of an application. Tasks can be submitted to run immediately, at a specified time, or in response to the completion of one or more actions upon which they depend.

For example, an application that computes the total cost of a merchandise order might look up the state and local sales tax rates in one task while calculating the shipping costs in another, and then combine the results in a third task that runs upon completion of the first two.

== Concurrency in Java SE and Java EE

Java SE has long included a https://docs.oracle.com/javase/8/docs/technotes/guides/concurrency/[Concurrency Utilities specification]. This specification standardizes an API for performing fundamental aspects of a concurrent programming model, such as:

* Creating threads
* Submitting tasks to run in parallel
* Scheduling tasks for execution at a future time
* Awaiting completion of tasks and obtaining the result of their execution

Java SE also includes various implementations, in some cases backed by thread pools, that enable varying degrees of extensibility and customization. These are well suited to the Java SE environment, where a single instance of an application services a single request and submits all of the associated tasks.

However, in a Java EE environment, multiple applications service multiple requests at any given moment.  These applications need to perform tasks concurrently in order to improve throughput and decrease response time. Rather than having each application maintain its own pool of threads, it is more efficient for applications to share thread pools and task scheduling.

_**should ManagedExecutorService be explained here? Its referenced in the next section but not mentioned here.**_

==Concurrency in MicroProfile

MicroProfile applications have many of the same needs as those in Java EE, but with a greater focus on reactive patterns.  One form of reactive programming is the Java SE [CompletionStage](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletionStage.html) model, which emphasizes creating pipelines of dependent actions that run upon completion of the states on which they depend.  However, in Java SE, these pipelines run with inconsistent thread contexts.  The [MicroProfile Context Propagation specification](https://download.eclipse.org/microprofile/microprofile-context-propagation-1.0/microprofile-context-propagation.html) solves this problem by introducing the [ManagedExecutor](https://openliberty.io/docs/ref/javadocs/microprofile-3.0-javadoc/org/eclipse/microprofile/context/ManagedExecutor.html), which is nearly identical to and fully compatible with the [ManagedExecutorService](https://openliberty.io/docs/ref/javadocs/liberty-javaee8-javadoc/javax/enterprise/concurrent/ManagedExecutorService.html) in Java EE, except that it is enhanced with the ability to create consistent, context-aware implementations of [CompletionStage](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletionStage.html). 

== Sharing thread pools

_**both thread factories and executors are referenced here, without being explained or defined, should they be?**_

By making executors and thread factories into managed resources (available in the Java namespace and injectable into applications), the https://docs.oracle.com/javaee/7/tutorial/concurrency-utilities.htm[Concurrency Utilities for Java EE specification] makes it possible for these executors and thread factories to be shared across applications and application components. Furthermore, Concurrency Utilities for Java EE introduces a standard set of default resources (for example, `java:comp/DefaultManagedScheduledExecutorService`) that are available to every application. This makes it extremely convenient for applications to share a single threading and scheduling implementation, managed by the application server, that requires no configuration. 

MicroProfile Context Propagation takes this a step further because the default managed executors all share a single common thread pool with the core server implementation. This thread pool is automatically tuned throughout the lifetime of the server by measuring the impact of incrementally increasing and decreasing the number of threads.

_**Is this a good spot to discuss context propagation?**_

== Commonalities between Java SE and EE concurrency and MicroProfile

The MicroProfile Context Propagation specification specification closely matches the Concurrency Utilities specifications for Java SE and Java EE. This commonality means you can write applications using Java SE or Java EE APIs and still take advantage of MicroProfile Context Propagation. You can even inject executor services and thread factories under the Java SE interface names (for example, `java.util.concurrent.ExecutorService`) as well as the Java EE interface names (for example, `javax.enterprise.concurrent.ManagedExecutorService`). With the [mpContextPropagation-1.0](https://openliberty.io/docs/ref/feature/#mpContextPropagation-1.0.html) feature enabled, you can also cast these executors to the `org.eclipse.microprofile.context.ManagedExecutor` interface, which enables context-aware completion stages.  Once you obtain the completion stage from the managed executor, you interact with it exactly as you would the Java SE API, but with the ability to rely on a consistent thread context in your completion stage actions.


== Thread context awareness

**_this section would benefit from a practrical application example, code snippet or both_**

In both Concurrency Utilities for Java EE and MicroProfile Context Propagation, executors are aware of thread context and can propagate the context of the task submitter to the thread of execution where the task or action runs. This allows tasks to perform lookups in the `java:comp`, `java:module`, or `java:app` namespace of the submitting thread, as well as to run with the security context of the submitting thread, load classes from its thread context classloader, and so on. With Liberty, you can also configure which types of thread context are propagated.

Thread factories are also aware of thread context and propagate the context of the thread that looked up or injected the thread factory to  the threads that factory creates. This enables thread pools that are created with these thread factories to create a pool of threads that all run under the same thread context, minimizing the costs of thread context switching. 

Managed executors and thread factories rely on another service provided by Concurrency Utilities for Java EE called the Context Service (`javax.enterprise.concurrent.ContextService`) for thread context propagation. Context Services are also available for direct usage, allowing you to achieve more flexible and fine grained control of thread context propagation. Context Services allow you to construct proxies that save the thread context at creation time and apply it upon invocation of interface methods. In MicroProfile Context Propagation, the [ThreadContext](https://openliberty.io/docs/ref/javadocs/microprofile-3.0-javadoc/org/eclipse/microprofile/context/ThreadContext.html) API provides similar function.

== Notifications and triggers

_**does MPCP provide similar notification and/ or trigger features?**_

Other advanced capabilities provided by Concurrency Utilities for Java EE are the ability to register for notifications of task lifecycle and a Trigger API that allows you to customize scheduling for complex business logic. A trigger allows you to recompute the next execution time each time the task runs. This means you are no longer limited to scheduling repeating tasks at fixed intervals and could, for example, schedule a task that only runs Monday through Friday, skipping weekend days. Or you could use Trigger to make the frequency of your task depend on other external stimuli, such as the weather or the rate at which sales are being made.

== Try out the Liberty Concurrency and MicroProfile Context Propagation features

**_what current or forthcoming guides should be linked here? (other than Nathan's MPCP topic, which is will cross link w/ this doc)_**

With the concurrent-1.0 feature, Liberty provides a robust, fully spec-compliant implementation of Concurrency Utilities for Java EE. MicroProfile Context Propagation is available with the mpContextPropagation-1.0 feature.  Both features can be enabled and used in tandem.  Try it out today, and start enjoying the advantages of parallelism and task scheduling in a Java EE environment.

**Comment by njr-11** [comment]
> should ManagedExecutorService be explained here? Its referenced in the next section but not mentioned here.

This section is an overview of some basic concepts, I don't think a definition of ManagedExecutorService belongs here, nor is it needed in the following section where a JavaDoc link is provided if the user cares to read more about it.  If you must include something, we could say that Java EE Concurrency extends the Java SE ExecutorService, ScheduledExecutorService, and ThreadFactory with subclasses of which the application server supplies the implementation: ManagedExecutorService, ManagedScheduledExecutorService, and ManagedThreadFactory.

> both thread factories and executors are referenced here, without being explained or defined, should they be?

Will the above change suffice to take care of this as well?  If we need to get into too much detail, this will cease being a good overview document.

> MicroProfile Context Propagation takes this a step further because the default managed executors all share a single common thread pool with the core server implementation.

Remove the words, "the default", because this statement applies to all managed executors, not only the built-in ones.

> Is this a good spot to discuss context propagation?

The later section on "Thread context awareness" covers the concept of context propagation.  Do you think we need anything else?

> this section would benefit from a practrical application example, code snippet or both

Here is an example of thread context propagation which should fit nicely after the first paragraph in this section

```
CompletableFuture<Long> stage = managedExecutor
    .supplyAsync(supplier)
    .thenApply(value -> {
        // only possible with the application component's context available:
        DataSource ds = InitialContext.doLookup("java:module/env/jdbc/ds1");
        ...
    });
```

> does MPCP provide similar notification and/ or trigger features?

no, which is why none are mentioned in this section

> what current or forthcoming guides should be linked here? (other than Nathan's MPCP topic, which is will cross link w/ this doc)

The [MP FaultTolerance Bulkhead guide](https://openliberty.io/guides/bulkhead.html) is being updated under https://github.com/OpenLiberty/guides-common/issues/219 to utilize MP Context Propagation.

**Comment by dmuelle** [comment]
@lauracowen I've incorporated @njr-11 edits and example and posted my draft to the draft site for review:

https://draft-openlibertyio.mybluemix.net/docs/ref/general/#context_propagation.html

Per Nathan I've focused on concurrency rather than context propagation per se, I should probably update the file name before it publishes. Let me know what other changes this needs.

One thing to note- I added in @njr-11 code snippet to the `Thread Context Awareness` section as suggested, but I think it needs an introductory sentence, which, unfortunately I can't derive from looking at the code ;(. For now I put in the placeholder `The following examples shows _____.` Nathan, can you fill in that blank or provide an intro?

Thanks!

**Comment by njr-11** [comment]
Please change:
"With MicroProfile Context Propagation, the default managed executors all share a single common thread pool..."
to:
"With MicroProfile Context Propagation, **all** managed executors all share a single common thread pool..."

"The following examples shows _"
can be:
"The following example shows usage of a managed executor that was configured to propagate the thread context type [ThreadContext.APPLICATION](https://openliberty.io/docs/ref/javadocs/microprofile-3.0-javadoc/org/eclipse/microprofile/context/ThreadContext.html#APPLICATION). This gives the second stage the ability to perform the `java:comp` lookup against the namespace of the application that invoked `thenApply`."

**Comment by lauracowen** [comment]
That's great, thank you both! I like how you've woven the MP CP information in with the EE and SE information and explained what each brings to an app with some nice examples. :)

Minor (really minor) things I spotted:
- "Triggers and Notifications" -> lower-case "notifications" (and check the other headings that they're sentence-style capitalisation too).
- "The following examples shows _." - too many "s" and random punctuation...
- I wouldn't use "for example" inside parentheses - I think "e.g." is fine - regardless of what IBM Style might say, I don't think there's a problem of comprehension using "e.g.". However, I'm not going to start an argument over it so do what you think is best. :)
- "Context Services are also available" and "With Context Services" - This seems like it should be lower-case "services" as it doesn't seem like it's referring to the name (which it was in the previous sentence) so much as the concept which doesn't need capitalising.
- Is it worth removing the version numbers "1.0" from the feature names in the final paragraph so we don't have to update it each time it changes? Alternatively, there might be an entity you can use - talk to Charlotte because I think she had something in her server topics that Alasdair asked her to use (it might be irrelevant to this but worth checking).

**Comment by dmuelle** [comment]
removed the link to concurrency-microservices.adoc until that's ready to publish

**Comment by dmuelle** [comment]
ok- made the following changes
- "Triggers and Notifications" -> lower-case "notifications" **DONE**
- "The following examples shows _." - too many "s" and random punctuation...**DONE** also added Nathan's explanation
- I wouldn't use "for example" inside parentheses - **DONE**
- "Context Services are also available" and "With Context Services" - **DONE**
- Is it worth removing the version numbers "1.0" from the feature names in the final paragraph so we don't have to update it each time it changes? **DONE** Figuring since we are currently at 1.0, there won't be confusion about which version and going forward, won't need an update
- updated file name
- fixed `all managed executors...` sentence 

I think this is ready to publish, opening a PR to dev for Laura to approve

**Comment by dmuelle** [comment]
doc is now live at https://www.openliberty.io/docs/ref/general/#concurrency.html

Need to restore link to concurrency_microservices when that doc is published, will note in #471

**Comment by lauracowen** [comment]
Nice! Thanks

**Comment by dmuelle** [comment]
Hi @lauracowen- with this doc published and links updated in correct format, is this issue clear to close?

**Comment by lauracowen** [comment]
Yes. Fine to close.
